src/module_4/notebook.md

```diff
@@ -0,0 +1,2102 @@
+    """
+    fpr, tpr, _ = roc_curve(y_test, y_pred)
+    roc_auc = auc(fpr, tpr)
+    precision_, recall_, _ = precision_recall_curve(y_test, y_pred)
+    pr_auc = auc(recall_, precision_)
+    return roc_auc, pr_auc
+```
+
+### Baseline
+
+We first try our baseline without any ML involved, just using a single column, `global_popularity` to predict the outcome.
+
+
+```python
+plot_roc_and_pr_curves("Baseline", y_pred=val_df["global_popularity"], y_test=val_df[label_col])
+baseline_auc = auc_score(val_df[label_col], val_df["global_popularity"])
+```
+
+
+    
+![png](notebook_files/notebook_18_0.png)
+    
+
+
+### Model training
+
+
+```python
+def feature_label_split(df: pd.DataFrame, label_col: str) -> Tuple[pd.DataFrame, pd.Series]:
+    X = df.drop(columns=[label_col], axis=1)
+    y = df[label_col]
+    return X, y
+
+X_train, y_train = feature_label_split(train_df, label_col)
+X_val, y_val = feature_label_split(val_df, label_col)
+X_test, y_test = feature_label_split(test_df, label_col)
+```
+
+
+```python
+train_cols = numerical_cols + binary_cols
+```
+
+### Linear Model
+
+From the task of the previous module, we plot the best models for reference. We need to make sure our non-linear models (more complex models) improve on performance in order to justify the increase in model complexity.
+
+
+```python
+pr_auc_score_val = []
+roc_auc_scores_val = []
+
+fig1, ax1 = plt.subplots(1, 2, figsize=(12, 5))
+fig1.suptitle("Train metrics")
+
+fig2, ax2 = plt.subplots(1, 2, figsize=(12, 5))
+fig2.suptitle("Validation metrics")
+
+lrs = [
+    make_pipeline(StandardScaler(), LogisticRegression(penalty='l2', C=1e-6)),
+    make_pipeline(StandardScaler(), LogisticRegression(penalty='l1', C=1e-4, solver='saga')),
+]
+
+names = ["Ridge 1e-6", "Lasso 1e-4"]
+for name, lr in zip(names, lrs):
+    lr.fit(X_train[train_cols], y_train)
+    train_proba = lr.predict_proba(X_train[train_cols])[:, 1]
+    plot_roc_and_pr_curves(name, y_pred=train_proba, y_test=y_train, figure=(fig1, ax1))
+    
+    val_proba = lr.predict_proba(X_val[train_cols])[:, 1]
+    plot_roc_and_pr_curves(name, y_pred=val_proba, y_test=y_val, figure=(fig2, ax2))
+    auc_scores_val = auc_score(y_val, val_proba)
+    roc_auc_scores_val.append(auc_scores_val[0])
+    pr_auc_score_val.append(auc_scores_val[1])
+    
+    coeffs = lr.named_steps["logisticregression"].coef_[0]
+    feat_importance = pd.DataFrame({"feature": train_cols, "importance": np.abs(coeffs)})
+    feat_importance = feat_importance.sort_values('importance', ascending=True)
+    feat_importance.plot(kind='barh', x='feature', y='importance', title="Feature Importance", figsize=(9, 5))
+    
+    
+plot_roc_and_pr_curves("Baseline", y_pred=val_df["global_popularity"], y_test=val_df[label_col], figure=(fig2, ax2))
+```
+
+
+    
+![png](notebook_files/notebook_24_0.png)
+    
+
+
+
+    
+![png](notebook_files/notebook_24_1.png)
+    
+
+
+
+    
+![png](notebook_files/notebook_24_2.png)
```

Comment:
Again, please let's be quite skeptical about this global_popularity feature. If it was calculated over the whole dataset it's clearly leaking